storage/imxrt-flash/flashsrv.c

```diff
@@ -12,78 +12,290 @@
+			odevctl->properties.ssize = memory->ctx.properties.sector_size;
+			odevctl->properties.offs = 0;
+			odevctl->err = EOK;
+			break;
+
+		case flashsrv_devctl_sync:
+			TRACE("imxrt-flashsrv: flashsrv_devctl_sync, id: %d, port: %d.", idevctl->oid.id, idevctl->oid.port);
+			flash_sync(&memory->ctx);
+			odevctl->err = EOK;
+			break;
+
+		case flashsrv_devctl_eraseSector:
+			TRACE("imxrt-flashsrv: flashsrv_devctl_eraseSector - adrr: %d, id: %d, port: %d.", idevctl->erase.addr, idevctl->oid.id, idevctl->oid.port);
+			flashsrv_eraseSector(memory->fOid.id, idevctl->erase.addr);
+			odevctl->err = EOK;
+			break;
 
-			odevctl->properties.fsize = flash_memory->context.properties.size;
-			odevctl->properties.psize = flash_memory->context.properties.page_size;
-			odevctl->properties.ssize = flash_memory->context.properties.sector_size;
+		case flashsrv_devctl_erasePartition:
+			TRACE("imxrt-flashsrv: flashsrv_devctl_eraseChip, id: %d, port: %d.", idevctl->oid.id, idevctl->oid.port);
+			flashsrv_chipErase(memory->fOid.id);
 			odevctl->err = EOK;
 			break;
 
+		default:
+			odevctl->err = -EINVAL;
+			break;
+	}
+}
+
+
+static void flashsrv_rawCtl(flash_memory_t *memory, msg_t *msg)
+{
+	int i;
+	uint32_t sNb;
+	uint8_t partID;
+	flash_i_devctl_t *idevctl = (flash_i_devctl_t *)msg->i.raw;
+	flash_o_devctl_t *odevctl = (flash_o_devctl_t *)msg->o.raw;
+
+	if (memory == NULL) {
+		odevctl->err = -EINVAL;
+		return;
+	}
+
+	switch (idevctl->type) {
+		case flashsrv_devctl_properties:
+			partID = idevctl->oid.id;
+			TRACE("imxrt-flashsrv: flashsrv_devctl_properties, id: %d, port: %d.", partID, idevctl->oid.port);
+			odevctl->properties.size = memory->parts[partID].sz;
+			odevctl->properties.psize = memory->ctx.properties.page_size;
+			odevctl->properties.ssize = memory->ctx.properties.sector_size;
+			odevctl->properties.offs = memory->parts[partID].offs;
+			break;
+
 		case flashsrv_devctl_sync:
 			TRACE("imxrt-flashsrv: flashsrv_devctl_sync, id: %d, port: %d.", idevctl->oid.id, idevctl->oid.port);
-			flash_memory = flashsrv_findFlash(idevctl->oid);
-			if (flash_memory == NULL) {
+			flash_sync(&memory->ctx);
+			odevctl->err = EOK;
+			break;
+
+		case flashsrv_devctl_eraseSector:
+			partID = idevctl->oid.id;
+			TRACE("imxrt-flashsrv: flashsrv_devctl_eraseSector - adrr: %d, id: %d, port: %d.", idevctl->erase.addr + memory->parts[partID].offs,
+					partID, idevctl->oid.port);
+
+			if (idevctl->erase.addr	> memory->parts[idevctl->oid.id].sz) {
 				odevctl->err = -EINVAL;
 				break;
 			}
 
-			flash_sync(&flash_memory->context);
+			flashsrv_eraseSector(memory->fOid.id, memory->parts[partID].offs + idevctl->erase.addr);
+			odevctl->err = EOK;
+			break;
+
+		case flashsrv_devctl_erasePartition:
+			TRACE("imxrt-flashsrv: flashsrv_devctl_erasePartition, id: %d, port: %d.", idevctl->oid.id, idevctl->oid.port);
+			partID = idevctl->oid.id;
+			sNb = memory->parts[partID].sz / memory->ctx.properties.sector_size;
```

Comment:
You use data from external source to address the memory without checking it's validity. It's an exploit.